### DISCUSSION TITLE
HowTo: Footer with CC Logo + Question about Language selector not working

### DISCUSSION THREAD

**Discussion by michaelkowalczyk** [LiaScript/LiaScript#176] [Q&A]
Hi, I have two questions 

1. howto add a constant footer which includes the Creative Commons Logo and licence information?
Example from a video:

![Screenshot - 03_06_2024 , 12_55_23](https://github.com/LiaScript/LiaScript/assets/90335938/9f3acef6-7829-44f8-aa66-e17edfd3f954)

2. I added a valid link to the English translation file in the document header. (Currently I just can work with .txt files on the server) 
However when I click to change the language, the file is not loaded. 
The Documentation on https://liascript.github.io/course/?https://raw.githubusercontent.com/liaScript/docs/master/README.md#109 does only relate to "Translate with Google (experimental)"

![Screenshot - 03_06_2024 , 12_57_28 003 002](https://github.com/LiaScript/LiaScript/assets/90335938/a236bffa-16e3-4719-ab49-e80e650e2248)

Thank you for any tips!

**Comment by andre-dietrich** [comment]
Hi @michaelkowalczyk ...

1. I think, modifying the footer will be tough, since this element changes according to the different modes. And overwriting this, will probably not work on very small devices ... however, you could use `attribute` for this purpose, which allows adding multiple lines/blocks, whereby you only have to take care of the correct indentation:

   ![image](https://github.com/LiaScript/LiaScript/assets/3089101/a5df6510-e9d2-4f12-bd50-9ff5dcb85848)

   [Link LiveEditor](https://liascript.github.io/LiveEditor/?/show/code/H4sIAAAAAAAAA2XOOwvCMBSG4f38ilNcdEiOKC5FBNvVTRcRkSSNbWibhFy8/HsV3Zw+nm961wVjIFIKRuakSzx0JuLdhR7f67McTOx0g9k2OqDAumbVEQejtI2acw6AeHUOpQj4Y3FSCuUTozhPu5R8LImyH5xo+N30ZtSNEdyFlj7yH5Fy4+hspNTlUZImvaK6vlRHtt9ejHKWx1tLy8XcP9jfz71tZwCMbQAmuPuWAbwAV21ettgAAAA=)

2. Sorry, I did not check this for the LiveEditor, but I can guarantee that if you publish your project on GitHub or somewhere else, this error will not appear at the LiaScript-course view... This problem is caused, since the preview in the LiveEditor has not a real URL that is substituted with the translation URL. However, the translation links should work even as relative or absolute URLs ... but I need to fix this in the preview :see_no_evil: ...

3. Textfiles is okay, LiaScript does not care about the URL at all, as long as it receives a document that contains Markdown content ...

**Reply by michaelkowalczyk** [reply]
Thanks for clarification
the Logo inside the Info Box seems like a possible solution :) Before I added it to the header logo.

Question about working with the local browser LiveEditor.
- When working localy in the browser with the liascript editor. Is there a folder on my harddrive which actually contains the .md file which I edit?
- if yes, could I use this folder to save the media files and use relative links in the script?
- currently the way I work is to copy my script into a txt hosted on a server and then view the it using  "load course" on https://liascript.github.io/  - also all media files are uploaded on the server and the permalinks are used in the script.

Regarding the language translation not loading
- Here is a demo file https://liascript.github.io/course/?https://api.allorigins.win/raw?url=https://euniwell.de.cool/wp-content/uploads/liascript/test/de.txt
- I added the -header script text- for reference, maybe I did forget to add something specific to make it work?
- The language change is still not working in Firefox or Chrome.

Another question regarding https://liascript.github.io/course/
- This is the local browser view right? 
- If yes, I find it a little bit confusing to see it on a url. Maybe a text info could help claryfying that it is local url view. 
- if no, how do I delete it from there?

![Screenshot - 04_06_2024 , 13_01_37 002](https://github.com/LiaScript/LiaScript/assets/90335938/0d45a799-1f5a-4860-88c5-4ce7935a86b9)

**Comment by andre-dietrich** [comment]
### Concerning the LiveEditor ...

No, there is no folder, on the hard drive, everything is stored within the browser. But you can upload images too, these will also be stored within the browser, only the name is replaced with a hash value.

![image](https://github.com/LiaScript/LiaScript/assets/3089101/15c29c12-6a3f-41c1-a26c-0d6d50d2fb32)

Using this on a tablet, you can also directly shoot an image or a video that gets uploaded into your browser.

You can then download the entire project as a zip-file

![image](https://github.com/LiaScript/LiaScript/assets/3089101/6befe61f-2678-4ec5-82b2-4ef12d4b5f8e)

This will include all images, videos, and sound files.

> Technically, all files and data is stored within the browsers IndexedDB as blob content.

If you need more control over the structure of your content, I would suggest installing VSCode, together with the preview plugin. There you can organize your content better. You can find a how-to here:

https://liascript.github.io/vscode/

### Regarding the language translation not loading

I see it not ... sorry for the confusion ... there was a feature request, that if on a large screen, the menu looses focus, then this menu will be closed. Clicking on the link will remove the focus and this happens, before the link is actually clicked. I will fix this in the next release, in the following setting this will work, but it will open the link and not add the liascript-URL to the beginning ... I will fix this too ... so far we had only used this with relative urls ...

![image](https://github.com/LiaScript/LiaScript/assets/3089101/1c7a1c95-2306-4c20-821f-237d1c60e24d)

### Another question regarding https://liascript.github.io/course/

Yes, these are all courses that have been loaded by the user, you can get there via the home-button. We do not monitor anything ... This behavior is intended, the browser will also cache this content, once it is loaded so users can always get back to previously visited courses. And if you install LiaScript on your mobile or Chrome, you can access all the content also offline.

![image](https://github.com/LiaScript/LiaScript/assets/3089101/83849723-5d04-4821-b1fb-f1bcfcdd1b29)

If you later, when your course is ready, switch to version 1.0.0 or above, all the states of quizzes, surveys, codes, tasks, will also be stored permanently within the browser (see section version for more information on this topic). This way we tried to guarantee that also remote users with a poor internet connectivity can reopen their courses, whenever they want...

If you want to delete a course, simply click onto the delete button on the card...

> What Information would you provide in this context?

### Additional Notes

If you have control over your wordpress installation, then you could enable CORS cross-origin-resource-sharing for the LiaScript website or ask your administrator.

https://biglinden.com/support/how-to-enable-cors-in-wordpress/

By allowing this, LiaScript will not use anymore the proxy-url https://api.allorigins.win/raw
but instead directly load your content. This proxy is just a fallback if the content cannot be loaded directly ...

**Reply by michaelkowalczyk** [reply]
Thank you for the fast and detailes reply Andre.

Media Upload

- I know about the upload feature. I did no utilize it because of the cryptic file naming after upload. 
- When trying to upload an icon and logo in the header the image was not shown
- Uploading a html file (h5p module) is currently not supported? I am embedding them from wordpress.

Language

- thank for looking into the problem
- you mentioned "relative urls ..." for the translation. How does that work when there is no actual folder where the other language .md files are stored localy? are the translations uploaded as well?

Regarding https://liascript.github.io/course/

- So anything anyone loads into https://liascript.github.io/ is then listed publicly on https://liascript.github.io/course/  and also can by deleted by anyone using the "Trash" icon?
- Its kind of an unxepected behaviour, not only from a privacy standpoint.
- If I would like to show somebody a preview of my course, without making it public, do I need to run a own liascript server?

CORS 

- Thanks for the info about howto improve accessibility. I did not realize that a 3rd party is involved to render the content.

**Comment by andre-dietrich** [comment]
Regarding: _So anything anyone loads into https://liascript.github.io/ is then listed publicly on https://liascript.github.io/course/ and also can by deleted by anyone using the "Trash" icon?_

No it is not publicly available, this is only shown and cached in your browser ... Simply open the website in another browser ... These cards are only shown if you have send this course to others, then their browser will store this course for later use as well... The LiaScript-Website and interpreter is only a static website, there is no other server and no tracking ...

Since It is a progressive web app it makes use of a lot of browser-based functionalities, but only on the browser at client-side... I hope the following link explains this better than I do ...

https://de.wikipedia.org/wiki/Progressive_Web_App#Offline-Funktionalit%C3%A4t

**Reply by michaelkowalczyk** [reply]
Perfect :)
I guess a text "this is only visible to you" would be sufficeint for clarification when accessing the urls.

**Reply by andre-dietrich** [reply]
Sure ... thanks for pointing this out ...

**Comment by andre-dietrich** [comment]
CORS ... sorry about this too ... no the proxy is only used, so that the LiaScript-Website at the client-browser is allowed to fetch the content... If you add the CORS settings or host your content on GitHub, then the LiaScript-Website will be able to directly load and render it ... so fetching and rendering happens also on the clients browser only ... there is no other entity in between ...

**Reply by michaelkowalczyk** [reply]
Understood. Thank you for patience and help understanding liascripts inner working ways and progressive web apps.

**Comment by andre-dietrich** [comment]
Hi, the translations file-problems should be solved in the new version ... simply reload your course or the live-editor ...

**Reply by michaelkowalczyk** [reply]
Yes, thank you for looking into it Andre. I confirm that language switching is working now.

Regarding relative path to translation files inside the header:

- is it the proper way of using absolute paths to load other language .md files?
OR
- how would you link to them in a relative way. The upload of a .md file into the project is not implemented, it is?

translation: English https://oer.com/en.txt
translation: French translations/french.md

**Reply by andre-dietrich** [reply]
You can use both, either absolute or relative. For the absolute case, the query or search parameter is simply replaced by the new absolute URL thus `https://liascript.github.io/course/?` remains, in this case also if the course has been packed to SCORM, the current URL will be used ... for the relative path, the URL of the current course is used as the base... 

I hope this makes sense ... in your case, you could use the absolute path, if you use nextcloud for hosting ...

And no, it is currently not possible to do this in the LiveEditor, but we will integrate a better file-navigation in the future, something that allows you to create more complex projects ....
